### PR TITLE
Add initial specification in Quint

### PR DESCRIPTION
**Pull request by dnkolegov**
This PR adds an initial formal specification of the ChonkyBFT protocol in Quint.
It also contains tests and experiments results we have performed.

### CODE REVIEW

**Review by pompon0 — APPROVED** [review]

File: spec/protocol-spec/README.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+ faulty replicas. All seven replicas are correct, though some of them may be slow.
+ This instance lets us test whether more correct replicas may introduce issues.
+
+ - [n6f1b1](./n6f1b1.qnt) is the protocol instance for `N=6`, `F=1`, and one
+ Byzantine replica. Five correct replicas should be able to tolerate one
+ Byzantine replica. This instance must be safe, e.g., `agreement_inv` must not
+ be violated.
+ 
+ - [n6f1b2](./n6f1b2.qnt) is the protocol instance for `N=6`, `F=1`, and two
+ Byzantines replica. Two Byzantine replicas have the critical mass to partition
```
Comment (konnov) [inline spec/protocol-spec/README.md:28]:
```suggestion
 Byzantine replicas. Two Byzantine replicas have the critical mass to partition
```

File: spec/protocol-spec/README.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+ the correct replicas. Some invariants such as `agreement_inv` are violated in
+ this instance.
+ 
+ - [n6f1b3](./n6f1b3.qnt) is the protocol instance for `N=6`, `F=1`, and three
+ Byzantines replica. Many invariants are violated in this instance very quickly.
```
Comment (konnov) [inline spec/protocol-spec/README.md:33]:
```suggestion
 Byzantines replicas. Many invariants are violated in this instance very quickly.
```

File: spec/protocol-spec/README.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+  ```
+  $ quint verify --random-transitions=true --invariant=phase_commit_example ./experiments/n6f1b0.qnt
+  ```
+
+## 3. State invariants
+
+In the course of writing the specification, we have developed a number of state
+invariants. By simulating and model checking the specification against these
+invariants increases our confidence in the correctness of the protocol.
```
Comment (konnov) [inline spec/protocol-spec/README.md:75]:
```suggestion
invariants, we have increased our confidence in the correctness of the protocol.
```

File: spec/protocol-spec/README.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+## 4. Checking state invariants
+
+**Randomized simulator.** Similar to finding examples, we run the randomized simulator
+to check correctness of the protocol against state invariants. For example:
+
+```sh
+quint run --max-steps=20 --max-samples=100000 --invariant=no_commit_equivocation_inv n6f1b3.qnt
+```
+
+By changing `--max-steps`, we give the simulator how deep the executions should be, e.g., up to 20 steps. By changing `--max-samples`, we give the simulator the number of executions to try, e.g., 100,000.
```
Comment (konnov) [inline spec/protocol-spec/README.md:108]:
```suggestion
By changing `--max-steps`, we tell the simulator how deep the executions should be, e.g., up to 20 steps. By changing `--max-samples`, we give the simulator the number of executions to try, e.g., 100,000.
```

File: spec/protocol-spec/experiments/agreement-violation.txt [hunk 5]
```diff
@@ -0,0 +1,202 @@
+[State 27]
```
Comment (konnov) [inline spec/protocol-spec/experiments/agreement-violation.txt:1]: This example is outdated. I propose to remove it from this PR.

File: spec/protocol-spec/experiments/n6f1b1_dead_lock.qnt [hunk 6]
```diff
@@ -0,0 +1,43 @@
+// A specification instance for n=6, f=1, and 1 Byzantine fault
+//
+/// Test a liveness issue where some validators have the HighQC but don't have the block payload and have to wait for it,
+/// while some other validators have the payload but don't have the HighQC and cannot finalize the block, and therefore
+/// don't gossip it, which causes a deadlock unless the one with the HighQC moves on and broadcasts what they have, which
+/// should cause the others to finalize the block and gossip the payload to them in turn.
```
Comment (konnov) [inline spec/protocol-spec/experiments/n6f1b1_dead_lock.qnt:6]:
```suggestion
/// should cause the others to finalize the block and gossip the payload to them in turn.
///
/// This test does not execute all steps, which shows that this concrete deadlock scenario is not feasible.
```

File: spec/protocol-spec/main.conf [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (konnov) [inline spec/protocol-spec/main.conf:1]: This is a temporary config. We do not have to merge it.

**Review by konnov — APPROVED** [review]
Looks great! I've left several comments.

File: spec/protocol-spec/main.conf (hunk 7 above)
Comment (dnkolegov) [reply spec/protocol-spec/main.conf:1]: ok

File: spec/protocol-spec/experiments/agreement-violation.txt (hunk 5 above)
Comment (dnkolegov) [reply spec/protocol-spec/experiments/agreement-violation.txt:1]: ok

**Review by konnov — APPROVED** [review]
Looks good to merge for me!

**Pull request merged**